### ISSUE TITLE
Failed IMU begin with error code -12

### ISSUE THREAD

**Issue by SorabhR** [tdk-invn-oss/motion.arduino.ICM42670P#14]
Hey,

I have DK-42670-P development board.
The spi pins are out on J4 for the development board. I have connected my esp32s3 spi pins there.
But I am unable to initialise the IMU it is failing with error code -12.
Please help if you have any idea to resolve this.

**Comment by tdk-opensource** [comment]
Thank you for opening this issue! We will look into it shortly.<br><br>Notifying @tdk-invn-oss/motion-maintainers @tdk-invn-oss/arduino-maintainers

**Comment by rbuisson-invn** [comment]
Hello @SorabhR,

Could you please you provide a trace of the SPI signals using scope or logic analyzer?
That would greatly help us providing support.

Regards,

**Comment by SorabhR** [comment]
I am sorry I don't have a logic analyser.

But I tried raw code of spi reading the who_am_i register of the imu and that is working fine.
So why is the initialisation only failing with this library?

**Comment by rbuisson-invn** [comment]
Could you please check the value you read for the who am i?

**Comment by SorabhR** [comment]
I got 0x67 for who_am_i

**Comment by rbuisson-invn** [comment]
Thanks, perfect so SPI signals integrity is OK.
Could you please share the code you use for the raw who am i read?

**Comment by SorabhR** [comment]
#include <SPI.h>

// Define pins
const 
int CS_PIN = 10;  // Chip select pin for SPI
const 

void setup() {
 
 // Initialize SPI
  
SPI.begin();

  
// Set CS pin as output
  
pinMode(CS_PIN, OUTPUT);
  digitalWrite(CS_PIN, HIGH); // Set CS high initially

  
Serial.begin(115200);

}


void loop() {
  
// Select the ICM-42670-P
  
digitalWrite(CS_PIN, LOW);

  // Send a command (example: read from a register, 0x75 is WHO_AM_I register)
  

SPI.transfer(0x75 | 0x80);  // Send WHO_AM_I command (read bit is 1)
  byte response = 

SPI.transfer(0x00);  // Read the response

  
// Deselect the ICM-42670-P
  

digitalWrite(CS_PIN, HIGH);

  

Serial.print("WHO_AM_I response: ");
  Serial.println(response, HEX);

  
delay(1000);
}

**Comment by rbuisson-invn** [comment]
You don't use Arduino SPI.beginTransaction() method ?
What mode is the SPI? (should be MSBFIRST, SPI_MODE3)

**Comment by SorabhR** [comment]
I have used the default spi mode

**Comment by rbuisson-invn** [comment]
Could you share the Arduino sketch source code ?
At least the part initializing the device.

**Comment by SorabhR** [comment]
I have used this code only exactly 
No changes this uses SPI MODE 0

**Comment by rbuisson-invn** [comment]
Do you call following lines somewhere?:
ICM42670 IMU(SPI,10);
IMU.begin();

(sorry for the terrible questions, but I need to reduce the scope to understand what could be the issue)

**Comment by SorabhR** [comment]
During the examples code of the library yes I do call it but in my raw code I do not

Infact IMU.begin only fails

**Comment by rbuisson-invn** [comment]
Do you confirm that you use 10 in ICM42670 IMU(SPI,**10**); ? (default is pin 8)

**Comment by SorabhR** [comment]
Yes I did that change

**Comment by rbuisson-invn** [comment]
Ok can you try to add the following line in your raw code before starting the who am i read (before digitalWrite(CS_PIN, LOW);):
SPI.beginTransaction(SPISettings(6000000, MSBFIRST, SPI_MODE3));

**Comment by SorabhR** [comment]
Okay I will try that and let you know 
Any other thing also can I try as I will try things after sometime

**Comment by rbuisson-invn** [comment]
Yes sure, try other valuesfor SPI clock frequecy and SPI_MODE parameters.

**Comment by SorabhR** [comment]
Does the library use SPI3 and 6MHz clock?

**Comment by rbuisson-invn** [comment]
Yes exactly!
SPI frequency can be specified, for instance:
IMU(SPI,10,1000000);

**Comment by SorabhR** [comment]
So as far as I understand library uses SPI mode 3 what if the sensor does not work on that and only works on spi mode 0?

**Comment by rbuisson-invn** [comment]
By default the ICM42670P supports both SPI modes 0 & 3:
![image](https://github.com/user-attachments/assets/1a9d5058-99f7-4a37-ab75-b3b704b90fd9)

**Comment by rbuisson-invn** [comment]
Another possible issue, other than SPI settings, could be the delay function:
    delayMicroseconds(us);
The library uses this function before reading the reset status, if this function is not correctly implemented on your platform, this read could occur too early.

**Comment by SorabhR** [comment]
I am using espcore3 in Arduino ide 2.1
If that helps you understand if delay function is implemented properly 

So spi mode 0 and mode 3 both should work.

**Comment by SorabhR** [comment]
Hey
The raw code is working with spi mode 3

**Comment by rbuisson-invn** [comment]
And with SPI clock @6MHz?

**Comment by SorabhR** [comment]
Yes that's working too

**Comment by rbuisson-invn** [comment]
Ok many thanks for these tests and feedback.
So it looks like SPI works as implemented in the library...

Let's try validating the delayMicroseconds....
Could you try the following code on your platform?
Serial.println(millis());
delayMicroseconds(1000);
Serial.println(millis());

And check that the increment between the 2 println is around 1000us.

Regards,

**Comment by SorabhR** [comment]
Yeah so the difference between the two millis is of 1 which is 1000us

**Comment by rbuisson-invn** [comment]
That's correct !
So everything works separately :)
Might be a stack size issue... Do you have any idea on how to increase stack size on your platform?

Alternatively, you could modify the Documents\Libraries\Arduino\ICM42670\src\ICM42670.cpp in function spi_read, something like following, to check every parameters:
static int spi_read(inv_imu_serif* serif, uint8_t reg, uint8_t * rbuffer, uint32_t rlen) {
  ICM42670* obj = (ICM42670*)serif->context;
  obj->spi->beginTransaction(SPISettings(obj->clk_freq, MSBFIRST, SPI_MODE3));
  digitalWrite(obj->spi_cs,LOW);
  obj->spi->transfer(reg | SPI_READ);
  obj->spi->transfer(rbuffer,rlen);
  digitalWrite(obj->spi_cs,HIGH);
  obj->spi->endTransaction();
Serial.println(obj->clk_freq);
Serial.println(rlen);
Serial.println(rbuffer[0]);
  return 0;
}

**Comment by SorabhR** [comment]
6000000
1
255
This is printed multiple times then
ICM42670 initialization failed: -12

**Comment by rbuisson-invn** [comment]
Could you please add Serial.println(obj->spi_cs); and Serial.println(reg);
The rbuffer[0]=0xFF means that MISO was not driven by the sensor.

**Comment by SorabhR** [comment]
10000000
1
103
10
117

Suddenly it started working but giving garbage values

**Comment by rbuisson-invn** [comment]
Hello @SorabhR ,
Your last reply shows a correct who am i read with SPI clock at 1M.
What's the garbage you are mentioning?

**Comment by SorabhR** [comment]
Hey
I change spi mode 0 in library and it started giving data

**Comment by rbuisson-invn** [comment]
Ok looks like SPI signal integrity is not so good finally.
Could you please try to write the DRIVE_CONFIG3 register to 0x3 or 0x2 ?
This will reduce the SPI slew rate, which is required to avoid crosstalk in some specific setup.
![image](https://github.com/user-attachments/assets/67967004-071e-469a-951c-d88b7bba22b9)

**Comment by SorabhR** [comment]
But with mode0 it is working fine so what would changing slew rate help in?

**Comment by rbuisson-invn** [comment]
If SPI signals are not fully safe you will probably have other issues later in your dev.
There should be no reason for mode 0 to work better than mode 3 (the only diff is the clock state when CS is high).

**Comment by SorabhR** [comment]
So In spi mode 0 I should change the slew rate but when I change the spimode3 will re powering both sensor and esp cause the settings to go back to factory default?

**Comment by rbuisson-invn** [comment]
The slew rate is reset at power up.
The register I pointed out changes the SPI signal form only for MISO.
So it doesn't impact write access.

**Comment by SorabhR** [comment]
So should I set that in the Imu.begin function?

**Comment by rbuisson-invn** [comment]
Yes before inv_imu_init, you could do something like:
uint8_t data = 0x03;
spi_write(&icm_serif,0x05,&data,1);

**Comment by SorabhR** [comment]
Okay

**Comment by rbuisson-invn** [comment]
Hello @SorabhR,

Any feedback on that test?

Regards,

**Comment by SorabhR** [comment]
Hey
Sorry for late reply but it didn't work only spi mode 0 is working

**Comment by kjs-eir** [comment]
@SorabhR and @rbuisson-invn any update on this issue? I am encountering the same issue, ( i have tried to change slew rate) and being trying both SPI4 and SPI3. 

The error code -12 happens to me after executing `status |= inv_imu_device_reset(s)` in `inv_imu_init()`
I don't get the issue when running of the eval kit, but once i move to my PCB with ICM42670 on it, I get the error code.

**Comment by rbuisson-invn** [comment]
Hello @kjs-eir,

Could you please share any capture of the SPI signals?

Thanks and regards,

**Comment by kjs-eir** [comment]
Hi @rbuisson-invn 

I have one setup with access to the SPI-bus, but this is working fine, and another setup on a PCB without access to the SPI-bus which isn't working(here i get -12 when trying to reset the IMU). I can share captures of the working SPI communication from the eval-kit, but does that make any sense? 

By the way, here is how my ICM42670P is wired on the PCB:
![image](https://github.com/user-attachments/assets/39bcec9d-1dd9-4fb0-beae-64234ed3a15e)


Thanks, 
Karl-Johan

**Comment by sriccardi-invn** [comment]
Hi @kjs-eir ,
Could you describe which PCBs you are using and connection between CPU and ICM42670?
Thanks.
Sebastien Riccardi

**Comment by kjs-eir** [comment]
Hi @sriccardi-invn 

Sure. For my evaluation board setup I use a Psoc6 MCU (Also on a custom PCB, but with way more test-points) and the DK-42670-P eval board. The connection is made via J4, where I connect MISO, MOSI, SCLK, CS, INT1, INT2, GND. I also connect a 3.3V source from my MCU. On DK-42670-P I have a jumper on J3 between 3V3 and VDDIO. This makes me able to power the DK-42670-P with out external power, and then I can power up the ICM42670P and get IMU via SPI. 

My second setup is a custom made PCB, but with the same MCU, and a ICM4670P. The connection are as shown on my previous picture, where SPI_MISO_IMU, SPI_MOSI_IMU, SPI_CLK_IMU, SPI_CS_IMU and IMU_INT are connected to the respective SPI-compatible GPIO on the MCU. IMU_WAKE is connected to a buck-boost converter, with the intension of using WOM. But here I encounter the issue with error code -12 when trying to reset the IMU.

I hope this make it easier to understand the issue, otherwise let me know.

Thanks, 
Karl-Johan

**Comment by sriccardi-invn** [comment]
HI @kjs-eir ,
Between both board, you are using same CPU io for SPI ?
thanks.
Regards,
Seb

**Comment by kjs-eir** [comment]
Hi @sriccardi-invn 

Yes, the same model and rev, both a PSOC6, and I am flashing the same FW image on both setup. 

Thanks, 
Karl-Johan

**Comment by sriccardi-invn** [comment]
HI @kjs-eir ,
Do you have a pull up on INT2 on your new board ? INT2 is open drain by default. Mayve you have no 3.3v on IMU VDDIO
Regards,
Seb

**Comment by kjs-eir** [comment]
Hi @sriccardi-invn 

I have no pull-up on neither of INT1 or INT2. I can try and add one to INT1 via the configurable HW blocks in PSOC6, but doing that on INT2 is more tricky. 
I have 3.3V on VDDIO,  I have verified that with my multimeter. 

Any other leads/ideas? 

Thanks,
Karl-Johan

**Comment by sriccardi-invn** [comment]
HI @kjs-eir ,
Could you confirm that you have no other slave on the same SPI bus?
This is maybe not the rootcause but leave INT2 floating at start up will create issue on your buck-boost converter. Maybe your buck-boost converter has an internal pull up ?
Regards,
Seb

**Comment by kjs-eir** [comment]
Hi @sriccardi-invn 
Yes, there is no other slaves on the SPI bus. 
The buck-boost converter has no internal pull-up, but can be enable by different sources, where INT2 is one of them, and I overlook this before, but there is a 100k pull-down on INT2 . 

Regards
Karl-Johan

**Comment by rbuisson-invn** [comment]
Hello @kjs-eir ,

Could you please try the following change in the ICP42670P library:
- in inv_imu_driver.c, in function configure_serial_interface, line 1486, add the following lines:
    value = 0x3;
    inv_imu_write_reg(s,DRIVE_CONFIG3, 1, &value);

![image](https://github.com/user-attachments/assets/3528d959-0b99-490f-b116-4d09c454e868)

**Comment by kjs-eir** [comment]
Hi @rbuisson-invn 

Thanks for the input. I tried to add the lines for updating DRIVE_CONFIG3, but the issue persist. To clarify a bit more, it is when executing `inv_imu_device_reset()` and when trying to read the value of INT_STATUS:

`/* Clear the reset done interrupt */
	status |= inv_imu_read_reg(s, INT_STATUS, 1, &data);
	if (data != INT_STATUS_RESET_DONE_INT_MASK)
		status |= INV_ERROR_UNEXPECTED;`
		
Data become 0xff, resulting in INV_ERROR_UNEXPECTED

**Comment by rbuisson-invn** [comment]
Hello @kjs-eir ,

Thanks for the feedback.
You are correct regarding the issue, this error usually happens when the SPI communication is not properly working (as this is the first check done in the driver).

The best way to proceed would be to get the SPI traces on the failing board, to understand what the issue could be.

Regards,

**Comment by kjs-eir** [comment]
@rbuisson-invn maybe this is a long shot. But the eval kit has a external pull-down on the FSYNC pin, but my setup doesn't have that and the datasheet says it ok to connect it directly to ground? 
![0](https://github.com/user-attachments/assets/43b58d37-c4e8-4d81-b5b7-bc0884e0452c)

**Comment by rbuisson-invn** [comment]
Hello,

This should not be a problem.

Regards,

**Comment by carbonadam** [comment]
@kjs-eir &  @rbuisson-invn Did you ever get it working in SPI mode?  Im facing similar issues on an esp32 c6.   I managed to get things working in i2c mode easily but I've tried all the things listed above and more and still can't get past the who am I returning -4

I don't have other slaves but I do have a SPI breakout connector that I have tested with other SPI devices that are working as expected.

![Image](https://github.com/user-attachments/assets/9ae2794f-7463-4410-aa15-38a00b24aa80)
![Image](https://github.com/user-attachments/assets/1e00eab5-b37e-42e3-9485-d38bad9c04a0)

I did some scope traces to see what things look like and Im not getting anything back at all from MISO.

![Image](https://github.com/user-attachments/assets/0b37fb8f-0d55-49c0-9439-0ab1dd25f228)
![Image](https://github.com/user-attachments/assets/aa633835-bcf7-406b-af56-0e1a0362e0e3)
![Image](https://github.com/user-attachments/assets/e378bbae-5543-4bde-adc0-f5e3e2bb2894)
![Image](https://github.com/user-attachments/assets/37f5ff97-8e74-4b2c-a15f-ccde54952afc)
